File: 43. GroupAnagrams.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
#include <algorithm>
using namespace std;

vector<vector<string>> groupAnagrams(vector<string>& strs) {
    unordered_map<string, vector<string>> groups;
    for (auto& str : strs) {
        string key = str;
        sort(key.begin(), key.end());
        groups[key].push_back(str);
    }
    vector<vector<string>> result;
    for (auto& p : groups) result.push_back(p.second);
    return result;
}

int main() {
    auto print = [](vector<vector<string>> res) {
        cout << "[";
        for (size_t i = 0; i < res.size(); i++) {
            if (i) cout << ", ";
            cout << "[";
            for (size_t j = 0; j < res[i].size(); j++) { if (j) cout << ", "; cout << res[i][j]; }
            cout << "]";
        }
        cout << "]\n";
    };
    vector<string> w1 = {"eat", "tea", "tan", "ate", "nat", "bat"};
    vector<string> w2 = {""};
    vector<string> w3 = {"a"};
    print(groupAnagrams(w1));
    print(groupAnagrams(w2));
    print(groupAnagrams(w3));
    return 0;
}

File: 43. GroupAnagrams.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class groupAnagrams {

    public static List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> groups = new HashMap<>();

        for (String str : strs) {
            char[] chars = str.toCharArray();
            Arrays.sort(chars);
            String key = new String(chars);

            groups.computeIfAbsent(key, k -> new ArrayList<>()).add(str);
        }

        return new ArrayList<>(groups.values());
    }

    public static void main(String[] args) {
        String[] words1 = {"eat", "tea", "tan", "ate", "nat", "bat"};
        String[] words2 = {""};
        String[] words3 = {"a"};

        System.out.println(groupAnagrams(words1));
        System.out.println(groupAnagrams(words2));
        System.out.println(groupAnagrams(words3));
    }
}
